### PR TITLE
Android: Use DialogFragment for AlertMessage

### PR DESCRIPTION
**Pull request by Ebola16**
Fixes https://bugs.dolphin-emu.org/issues/7988
Fixes https://bugs.dolphin-emu.org/issues/10770

As far as I can tell, this works (even with multiple panic alerts in quick succession) but there are some things that I'm not sure I've done properly:

- ~~Did I properly handle AlertMessages while booting the core?~~
- ~~Since `NativeLibrary.displayAlertMsg` has a lock object, pausing emulation while waiting on the lock would freeze AlertMessages after rotating the screen. As far as I can tell, pausing emulation here seems unnecessary but correct me if I'm wrong.~~

Test case was https://bugs.dolphin-emu.org/issues/11424

### CODE REVIEW

**Comment by JosJuice** [comment]
> Since `NativeLibrary.displayAlertMsg` has a lock object, pausing emulation while waiting on the lock would freeze AlertMessages after rotating the screen. As far as I can tell, pausing emulation here seems unnecessary but correct me if I'm wrong.

Yes, not pausing emulation is probably the best workaround. This is the same issue as with PR #8819.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      setCancelable(false);
+      builder.setPositiveButton(android.R.string.ok, (dialog, which) ->
+      {
+        dialog.dismiss();
+        NativeLibrary.NotifyLock();
+      });
+    }
+    else
+    {
+      sAlertResult = false;
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java:61]: Is there a reason for this line?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/NativeLibrary.java [hunk 2]
```diff
@@ -463,74 +466,52 @@ public static boolean displayAlertMsg(final String caption, final String text,
...
+  {
+    return sAlertMessageText;
+  }
+
+  public static boolean IsAlertMessage()
+  {
+    return sIsAlertMessage;
+  }
+
+  public static void NotifyLock()
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/NativeLibrary.java:507]: Maybe it would make sense to specify what this lock is for, like `NotifyAlertMessageLock`. (Same with the actual variable name.)

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java (hunk 1 above)
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java:61]: I think if a panic alert sets `sAlertResult` to true, a later panic alert will still return sAlertResult = true if a user backs out of the `yesNo` prompt.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java [hunk 3]
```diff
@@ -389,7 +392,22 @@ public synchronized void clearSurface()
           NativeLibrary.WaitUntilDoneBooting();
         }
 
-        NativeLibrary.SurfaceDestroyed();
+        // Handle AlertMessages while booting the core.
+        if (NativeLibrary.IsBooting())
+        {
+          if (NativeLibrary.GetAlertMessageText() != null)
+          {
+            throw new IllegalStateException(NativeLibrary.GetAlertMessageText());
```
[the remark this answers is not in the document]
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:400]: Yes, I'm not sure the best way to handle this. The choices I thought of were deadlocking for an unknown reason or crashing for a known reason. I also assume panic alerts during core booting indicate major emulation problems anyway. This does fix `AlertMessages` outside of core booting so I think it's a worthwhile change that isn't worse than what we currently have. I'm open to other suggestions though.

**Comment by Ebola16** [comment]
@dolphin-emu-bot rebuild

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java (hunk 1 above)
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java:61]: But doesn't every code path that calls `NotifyAlertMessageLock` set `sAlertResult`? So either there is some way to dismiss the alert without calling `NotifyAlertMessageLock`, or there shouldn't be any need for this line.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:400]: I guess the other option would be to dismiss the dialog if you try to rotate the screen while showing an alert message while booting. While it does avoid crashing, it is a bit odd, so I'm not sure if it's actually better...

I guess either a crash or what I mentioned would be okay, but you should leave a comment about it in either case (either about this crash being intentional, or about why we dismiss the dialog).

File: Source/Core/Core/Core.cpp [hunk 4]
```diff
@@ -170,6 +170,11 @@ void DisplayMessage(std::string message, int time_in_ms)
   OSD::AddMessage(std::move(message), time_in_ms);
 }
 
+bool IsBooting()
+{
+  return (s_is_booting.IsSet() || !s_hardware_initialized);
```
Comment (JosJuice) [inline Source/Core/Core/Core.cpp:175]:
```suggestion
  return s_is_booting.IsSet() || !s_hardware_initialized;
```

Also, `WaitUntilDoneBooting` should be made to call this function.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:400]: Also, if you're going to have it crash, it probably makes more sense to always include "Attempted to destroy surface while booting the core" in the exception message (perhaps with the alert message text appended), since the error that causes the alert message to be shown is not in itself the reason why we are crashing.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java (hunk 1 above)
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java]: It turns out directly copying the old code into the DialogFragment didn't handle the case of backing out of a `yesNo` prompt, which never notified the lock. I removed the extra `sAlertResult` and made `yesNo` prompts not cancelable too. This makes sense since users can rapidly tap and accidentally dismiss the prompt.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: I tried another idea. Instead of deadlocking or crashing the emulator, we can simply end emulation. The AlertMessage has been replaced with a log error and toast. Hopefully this won't cause unforeseen problems.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java (hunk 1 above)
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java]: Sounds good to me.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: I also moved the `R.string.emulation_touch_button_help` toast to only run once during emulation. This will prevent it from reappearing when other toasts are displayed.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java [hunk 5]
```diff
@@ -389,7 +392,32 @@ public synchronized void clearSurface()
...
+            Toast.makeText(context, alertMessageError, Toast.LENGTH_LONG).show();
+          }
+          else
+          {
+            String generalError = "Attempted to destroy surface while booting the core.";
+
+            Log.error("[EmulationFragment] " + generalError);
+            Toast.makeText(context, generalError, Toast.LENGTH_LONG).show();
+          }
+          stop();
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:414]: Are you sure that calling `StopEmulation` while booting works correctly? This seems very fishy to me.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java [hunk 6]
```diff
@@ -389,7 +392,32 @@ public synchronized void clearSurface()
           NativeLibrary.WaitUntilDoneBooting();
         }
 
-        NativeLibrary.SurfaceDestroyed();
+        // AlertMessages while booting the core will dereference nullptr. Silently dismissing
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:395]: I don't understand how that would lead to a nullptr dereference. Could you clarify the comment with e.g. where we get the nullptr dereference?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: Ending emulation is also kinda weird, but I guess it's no worse than crashing.

What about just turning the alert message into a toast and then calling `WaitUntilDoneBooting` as usual, without stopping emulation or anything?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 6 above)
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:395]: If you're referring to the same nullptr dereference as in the comment above, you should write that we can't call `WaitUntilDoneBooting` if we get an alert message while booting rather than just writing that we will dereference nullptr.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: If `WaitUntilDoneBooting` is called instead of stopping emulation, we'll get the deadlock. I think that is the least appealing option of deadlock vs. crash vs. ending emulation. Still debating if I should go back to crashing the emulatior as I'm also not certain of the safety of stopping emulation either.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 6 above)
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java:395]: Yes, I was referring to the same nullptr dereference.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: > If `WaitUntilDoneBooting` is called instead of stopping emulation, we'll get the deadlock.

Yes, unless we release the lock first. My suggestion is that we dismiss the dialog and release the lock, and then show a toast.

[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: I just realized that regardless of what we do about this case, we're still going to get a deadlock in the case where we get an alert message after `WaitUntilDoneBooting` already has been called... I wonder if we maybe just should show all alert messages that are triggered when booting as toasts without even attempting to use a proper dialog.

[the remark this answers is not in the document]
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: I could use the new `NativeLibrary.IsBooting` check to handle alert messages differently during core boot. Just to clarify, this PR doesn't deadlock with alert messages after core boot because it skips calling `WaitUntilDoneBooting` for `AlertMessages`.

[the remark this answers is not in the document]
Comment (JosJuice) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: Yes, there is no problem after we're done booting.

[the remark this answers is not in the document]
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/fragments/EmulationFragment.java]: Moving the implementation of `IsBooting` and making alert messages toasts while the core is booting seems to be a good solution here. Hopefully this version does the trick!

**Comment by Ebola16** [comment]
Thanks for the help! This PR went in directions I didn't expect.

**Comment by Ebola16** [comment]
Minor typo fix. I accidentally called the "AlertMessage" a "Panic Alert" in `NativeLibrary`.

**Comment by Ebola16** [comment]
Can this be merged?

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java [hunk 7]
```diff
@@ -71,6 +75,17 @@ public Dialog onCreateDialog(Bundle savedInstanceState)
                 NativeLibrary.NotifyAlertMessageLock();
               });
     }
+
+    if (isWarning)
+    {
+      builder.setNeutralButton("IGNORE FOR THIS SESSION", (dialog, which) ->
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java:81]: This string should be in a resources file and should not be in allcaps.

File: Source/Android/jni/MainAndroid.cpp [hunk 8]
```diff
@@ -146,14 +146,14 @@ void Host_TitleChanged()
...
-static bool MsgAlert(const char* caption, const char* text, bool yes_no, Common::MsgType /*style*/)
+static bool MsgAlert(const char* caption, const char* text, bool yes_no, Common::MsgType style)
 {
   JNIEnv* env = IDCache::GetEnvForThread();
 
   // Execute the Java method.
   jboolean result = env->CallStaticBooleanMethod(
       IDCache::GetNativeLibraryClass(), IDCache::GetDisplayAlertMsg(), ToJString(env, caption),
-      ToJString(env, text), yes_no ? JNI_TRUE : JNI_FALSE);
+      ToJString(env, text), yes_no ? JNI_TRUE : JNI_FALSE, style == Common::MsgType::Warning);
```
Comment (JosJuice) [inline Source/Android/jni/MainAndroid.cpp:156]: I don't know if having the style affect the behavior of the panic alert is a good idea or not. It would be better if this commit could be split into another PR so that the merge of the other fixes in this PR aren't blocked by this question being open.

Comment (JosJuice) [reply Source/Android/jni/MainAndroid.cpp:156]: Actually nevermind lol we already do this in Qt. So I guess it's fine

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/NativeLibrary.java [hunk 9]
```diff
@@ -478,76 +481,64 @@ public static boolean displayAlertMsg(final String caption, final String text,
     {
       Log.warning("[NativeLibrary] EmulationActivity is null, can't do panic alert.");
     }
+    else if (emulationActivity.isIgnoringWarnings() && isWarning)
+    {
+      return false;
```
Comment (JosJuice) [inline Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/NativeLibrary.java:486]: To be consistent with DolphinQt, this should return true.

File: Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java (hunk 7 above)
Comment (Ebola16) [reply Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/dialogs/AlertMessage.java]: The string is converted to all caps when displayed but I now use a string resource that matches Qt.

**Comment by JosJuice** [comment]
Well, it seems like nobody else is going to review this...
